causallib/time_varying/GFormula.py

```diff
@@ -0,0 +1,225 @@
+#!/usr/bin/env python3
+
+import abc
+import pandas as pd
+from typing import Optional, Any, OrderedDict
+from causallib.time_varying.base_GFormula import TimeVaryingBaseEstimator
+from causallib.utils import general_tools as g_tools
+
+
+class GFormulaBase(TimeVaryingBaseEstimator):
+    """
+        GFormula base Estimator
+    """
+    def __init__(self,
+                 outcome_model: None,
+                 treatment_model: Any,
+                 covariate_models: OrderedDict,
+                 refit_models=True):
+        """
+            outcome_model(IndividualOutcomeEstimator): A causal model that estimate on individuals level
+                                                      (e.g. Standardization).
+            treatment_model (???):
+            covariate_models (OrderedDict): {
+                        ‘x1’: Standardization(LogisticRegression()),
+                        ‘x2’: Standardization(LinearRegression()),
+                            so forth
+                        },
+            refit_models (bool): if True, re-fit a the treatment model and covariate models.
+
+        """
+        super(GFormulaBase, self).__init__(lambda **x: None)
+        self.outcome_model = outcome_model
+        self.treatment_model = treatment_model
+        self.covariate_models = covariate_models
+        self.refit_models = refit_models
+
+
+    @abc.abstractmethod
+    def fit(self,
+            X: pd.DataFrame,
+            a: pd.Series,
+            t: pd.Series,
+            y: Optional[Any] = None,
+            refit_models: bool = True,
+            **kwargs
+            ):
+        """
+            Fits parametric models and calculates prediction curves.
+
+            Args:
+                X (pd.DataFrame): Baseline covariate matrix of size (num_subjects, num_features).
+                a (pd.Series): Treatment assignment of size (num_subjects,).
+                t (pd.Series): Followup duration, size (num_subjects,).
+                y (pd.Series): Observed outcome (1) or right censoring event (0), size (num_subjects,).
+                refit_models (bool): if True, re-fit a the treatment model and covariate models.
+                kwargs (dict): Optional kwargs for fit call of survival model
+
+            Returns:
+                self
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def estimate_individual_outcome(self,
+                                    X: pd.DataFrame,
+                                    a: pd.Series,
+                                    t: pd.Series,
+                                    y: Optional[Any] = None,
+                                    timeline_start: Optional[int] = None,
+                                    timeline_end: Optional[int] = None
+                                    ) -> pd.DataFrame:
```

Comment:
Add a variable for treatment strategy. Possible solution:

```
treatment_strategy: Callable,
```

Reply:
I thought the strategy could be defined according to the provided `a`. Can we discuss it on our next call?

Reply:
sure lets do that. `a` can capture the historical treatments but it needs a way to get the function to get the next one